dotnet/sdk-phase-2.md

```diff
@@ -0,0 +1,484 @@
+
+    public Task SignalAsync(
+        Func<Task> signal, ExternalWorkflowSignalOptions? options = null) { /*...*/ }
+
+    public Task SignalAsync<T>(
+        Func<T, Task> signal, T arg, ExternalWorkflowSignalOptions? options = null) { /*...*/ }
+
+    public Task SignalAsync<T>(
+        string signal, object[] args, ExternalWorkflowSignalOptions? options = null) { /*...*/ }
+}
+
+public class ContinueAsNewException { /*...*/ }
+```
+
+Notes about the above code:
+
+* Activities and child workflows can be provided cancellation tokens in their options
+  * Technically we could put `CancellationTokenSource` on the handles, but the current approach seems more .NET like
+* Since activities have no reason for an `ActivityHandle`, there is no `StartActivity`
+  * Python has `start_activity` + `ActivityHandle` because that language allows you to cancel the task
+  * The cancellation token in the options and the result of the task are enough to support all current activity features
+  * `StartActivity` + `ActivityHandle` may be added in the future if we ever allow doing things to activities more than
+    just cancelling and getting result
+* Signal/query handlers
+  * We could have made dynamic signal/query handler get/set a property but it's more consistent with non-dynamic to use
+    methods
+  * We could have returned/accepted null for unset, but this approach is more .NET like
+  * We currently have no public way to get all signal/query handlers from inside a workflow (same as Python)
+* There's no value in accepting workflow type in `GetExternalWorkflowHandle` at this time because we can only signal and
+  cancel it currently. And signal is not currently scoped to the workflow at compile time.
+* We expect users to `throw new ContinueAsNewException()` instead of have some helper
+  * This is the .NET preferred approach (same for `throw new CompleteAsyncException()` for activities)
+* We may add helpers for getting individual memo values given a type to deserialize to
+  * We don't expose a data converter to the user in a workflow at this time, so it is wise to make these helpers
+* A `CancellationToken` can be provided on every async call made, but if not set, by default workflow cancellation token
+  is the token used. This allows it to be overridden with a token not related to workflow cancellation.
+
+TODO(cretz): For discussion:
+
+* `ChildWorkflowHandle` has `GetResultAsync()` to match `WorkflowHandle` in client, but this could just be a `Task`
+  property. Should it be? Other .NET APIs tend to do this.
+* In client I call options `WorkflowStartOptions`, not only for `StartWorkflowAsync` but also for
+  `ExecuteWorkflowAsync`. I have a similar pattern here with activities and children. Should I continue this or should I
+  remove the `Start` word?
+* Should I rename `Now` to `UtcNow` to match .NET since it's non-local? Or does matching `DateTime.UtcNow` not matter
```

Comment:
I'd vote for matching .NET here and trading off verbosity for clarity.

Reply:
It's less about verbosity and more about that "now" is just a clearer term across our SDKs. But I can change to `UtcNow` which may be a tad less clear/discoverable but that should be ok.

Reply:
Fixed